### PR TITLE
make all component SSR compatible

### LINKED ISSUE #9111: Review components for SSR compatibility
In order to implement #8795 and for general usability of our components as standalone npm packages, we need to ensure that all component are SSR compatible.

**Components**

- [ ] [Accordion](https://www.gradio.app/docs/gradio/accordion)
- [ ] [Column](https://www.gradio.app/docs/gradio/column)
- [ ] [Group](https://www.gradio.app/docs/gradio/group)
- [ ] [Row](https://www.gradio.app/docs/gradio/row)
- [ ] [Tab](https://www.gradio.app/docs/gradio/tab)
- [ ] [AnnotatedImage](https://www.gradio.app/docs/gradio/annotatedimage)
- [ ] [Audio](https://www.gradio.app/docs/gradio/audio)
- [ ] [BarPlot](https://www.gradio.app/docs/gradio/barplot)
- [ ] [Button](https://www.gradio.app/docs/gradio/button)
- [ ] [Chatbot](https://www.gradio.app/docs/gradio/chatbot)
- [ ] [Checkbox](https://www.gradio.app/docs/gradio/checkbox)
- [ ] [CheckboxGroup](https://www.gradio.app/docs/gradio/checkboxgroup)
- [ ] [ClearButton](https://www.gradio.app/docs/gradio/clearbutton)
- [ ] [Code](https://www.gradio.app/docs/gradio/code)
- [ ] [ColorPicker](https://www.gradio.app/docs/gradio/colorpicker)
- [ ] [Dataframe](https://www.gradio.app/docs/gradio/dataframe)
- [ ] [Dataset](https://www.gradio.app/docs/gradio/dataset)
- [ ] [DateTime](https://www.gradio.app/docs/gradio/datetime)
- [ ] [DownloadButton](https://www.gradio.app/docs/gradio/downloadbutton)
- [ ] [Dropdown](https://www.gradio.app/docs/gradio/dropdown)
- [ ] [DuplicateButton](https://www.gradio.app/docs/gradio/duplicatebutton)
- [ ] [File](https://www.gradio.app/docs/gradio/file)
- [ ] [FileExplorer](https://www.gradio.app/docs/gradio/fileexplorer)
- [ ] [Gallery](https://www.gradio.app/docs/gradio/gallery)
- [ ] [HighlightedText](https://www.gradio.app/docs/gradio/highlightedtext)
- [ ] [HTML](https://www.gradio.app/docs/gradio/html)
- [ ] [Image](https://www.gradio.app/docs/gradio/image)
- [ ] [ImageEditor](https://www.gradio.app/docs/gradio/imageeditor)
- [ ] [JSON](https://www.gradio.app/docs/gradio/json)
- [ ] [Label](https://www.gradio.app/docs/gradio/label)
- [ ] [LinePlot](https://www.gradio.app/docs/gradio/lineplot)
- [ ] [LoginButton](https://www.gradio.app/docs/gradio/loginbutton)
- [ ] [LogoutButton](https://www.gradio.app/docs/gradio/logoutbutton)
- [ ] [Markdown](https://www.gradio.app/docs/gradio/markdown)
- [ ] [Model3D](https://www.gradio.app/docs/gradio/model3d)
- [ ] [MultimodalTextbox](https://www.gradio.app/docs/gradio/multimodaltextbox)
- [ ] [Number](https://www.gradio.app/docs/gradio/number)
- [ ] [ParamViewer](https://www.gradio.app/docs/gradio/paramviewer)
- [ ] [Plot](https://www.gradio.app/docs/gradio/plot)
- [ ] [Radio](https://www.gradio.app/docs/gradio/radio)
- [ ] [ScatterPlot](https://www.gradio.app/docs/gradio/scatterplot)
- [ ] [SimpleImage](https://www.gradio.app/docs/gradio/simpleimage)
- [ ] [Slider](https://www.gradio.app/docs/gradio/slider)
- [ ] [State](https://www.gradio.app/docs/gradio/state)
- [ ] [Textbox](https://www.gradio.app/docs/gradio/textbox)
- [ ] [Timer](https://www.gradio.app/docs/gradio/timer)
- [ ] [UploadButton](https://www.gradio.app/docs/gradio/uploadbutton)
- [ ] [Video](https://www.gradio.app/docs/gradio/video)

### LINKED ISSUE #9112: Make all incompatible components SSR compatible
Continuation of #9111.

This is to fix them.

### PR DESCRIPTION
**Pull request by pngwn**
**description**

closes #9111 closes #9112

Makes all component SSR compatible.

- Removes any bare references to browser only APIs, or hide them behind on Mount hooks or actions we know can only occur when we are in the browser and fully loaded
- Stop using location.origin and instead use the root url value that we have access to.
- Added some sensible defaults to avoid things jumping around too much when hydrated in the browser
- Use empty value for certain components until we are in the browser, things like native plots or model 3d don't make much sense in a server context. We can improve this later if needed.
- Added an SSR-able test app with some example values.
- Switch to an isomophic (server + browser compatible) version of `dompurify`.

**to test**

```
pnpm i
cd js/component-test
pnpm dev
```

While running the app. I would recommend doping the following to see how things behave:

- Visit every link at the top and hard refresh the browser (with dev tools closed to you get cached fonts). This is Server Side Rendering + Client Side Rendering (hydration).
- Try with JS disabled:
   - Go into chrome dev tools
   - Click the cog in the top right 
   - scroll down to "Debugger"
   - Click "Disable JavaScript"
   - Keep dev tools open and revisit all links + refresh, this is SSR only.

### CODE REVIEW

File: js/imageeditor/shared/ImageEditor.svelte [hunk 1]
```diff
@@ -84,13 +84,13 @@
 
 	let editor_box: EditorContext["editor_box"] = writable({
 		parent_width: 0,
-		parent_height: 0,
+		parent_height: 400,
```
Comment (pngwn) [inline js/imageeditor/shared/ImageEditor.svelte:87]: use provided height value if it is passed

File: js/markdown/package.json [hunk 2]
```diff
@@ -31,6 +31,7 @@
 		"@types/prismjs": "1.26.4",
 		"dompurify": "^3.0.3",
```
Comment (pngwn) [inline js/markdown/package.json:32]: remove

File: js/markdown/shared/MarkdownCode.svelte [hunk 3]
```diff
@@ -74,7 +66,15 @@
...
-
+		DOMPurify.addHook("afterSanitizeAttributes", function (node) {
+			if ("target" in node) {
+				if (is_external_url(node.getAttribute("href"))) {
+					node.setAttribute("target", "_blank");
+					node.setAttribute("rel", "noopener noreferrer");
+				}
+			}
+		});
+		const is_browser = typeof window !== "undefined";
```
Comment (pngwn) [inline js/markdown/shared/MarkdownCode.svelte:77]: Does this need to be moved?

Do we need the browser var?

File: js/wasm/svelte/DownloadLink.svelte [hunk 4]
```diff
@@ -64,9 +64,14 @@
 				is_downloading = false;
 			});
 	}
+
+	let should_proxy = false;
+	onMount(() => {
```
Comment (pngwn) [inline js/wasm/svelte/DownloadLink.svelte:69]: Remove?

**Comment by abidlabs** [comment]
>  @pngwn was reviewing the components yesterday and it looked good to me. But now I'm getting "SSR Module" errors when running the demo. Could just be me tho.

Locally or on the Deployed PR Space @dawoodkhan82?

**Comment by dawoodkhan82** [comment]
@abidlabs locally

**Comment by abidlabs** [comment]
I can't repro that @dawoodkhan82. I checked out this branch, rebuilt the frontend, and ran some demos (like `kitchen_sink`) and they worked fine for me. I do have Node v18 installed in case that makes any difference

**Comment by dawoodkhan82** [comment]
@abidlabs did you try with disabled JS from the dev tools?

**Comment by abidlabs** [comment]
Ah yeah sorry missed those testing instructions. I'm seeing 500 errors as well. The web app looks like this:

image

Stack trace:

```
1:36:25 PM [vite] Error when evaluating SSR module /@fs/Users/abidlabs/dev/gradio-repos/gradio/js/atoms/src/ShareButton.svelte: failed to import "@gradio/utils"
|- Error: Failed to resolve entry for package "@gradio/utils". The package may have incorrect main/module/exports specified in its package.json.
    at packageEntryFailure (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:48264:17)
    at resolvePackageEntry (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:48261:5)
    at tryNodeResolve (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:48039:20)
    at nodeImport (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55087:26)
    at ssrImport (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:54998:30)
    at eval (/Users/abidlabs/dev/gradio-repos/gradio/js/atoms/src/ShareButton.svelte:7:37)
    at async instantiateModule (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55058:9)

1:36:25 PM [vite] Error when evaluating SSR module /@fs/Users/abidlabs/dev/gradio-repos/gradio/js/atoms/src/index.ts: failed to import "/@fs/Users/abidlabs/dev/gradio-repos/gradio/js/atoms/src/ShareButton.svelte"
|- Error: Failed to resolve entry for package "@gradio/utils". The package may have incorrect main/module/exports specified in its package.json.
    at packageEntryFailure (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:48264:17)
    at resolvePackageEntry (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:48261:5)
    at tryNodeResolve (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:48039:20)
    at nodeImport (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55087:26)
    at ssrImport (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:54998:30)
    at eval (/Users/abidlabs/dev/gradio-repos/gradio/js/atoms/src/ShareButton.svelte:7:37)
    at async instantiateModule (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55058:9)

1:36:25 PM [vite] Error when evaluating SSR module /@fs/Users/abidlabs/dev/gradio-repos/gradio/js/multimodaltextbox/shared/MultimodalTextbox.svelte: failed to import "/@fs/Users/abidlabs/dev/gradio-repos/gradio/js/atoms/src/index.ts"
|- Error: Failed to resolve entry for package "@gradio/utils". The package may have incorrect main/module/exports specified in its package.json.
    at packageEntryFailure (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:48264:17)
    at resolvePackageEntry (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:48261:5)
    at tryNodeResolve (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:48039:20)
    at nodeImport (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:55087:26)
    at ssrImport (file:///Users/abidlabs/dev/gradio-repos/gradio/node_modules/.pnpm/vite@5.2.11_@types+node@20.12.8_lightningcss@1.24.1_sass@1.66.1_stylus@0.63.0_sugarss@4.0.1_postcss@8.4.38_/node_modules/vite/dist/node/chunks/dep-cNe07EU9.js:54998:30)
```

**Comment by pngwn** [comment]
I'll take a look at this. Seems to just be the utils package that is causing issues.

**Comment by pngwn** [comment]
@abidlabs @dawoodkhan82 I left something out of my instructions which doesn't affect other packages but apparently does affect the utils package for some reason.

You need to run `pnpm packages` from the root first. That should fix the issue you were seeing.

**Comment by abidlabs** [comment]
Ok working for me now, but noticing a few small things. Let me comment as I find them:

(1) i18n strings are not rendering correctly, e.g. 

image

image

(2) Audio player is not coloring the rendered audio appropriately. You can see here that its hiding the played audio instead of coloring it orange:

https://github.com/user-attachments/assets/51f512d0-cb6b-4168-aba6-c5e19f71a5d3

**Comment by abidlabs** [comment]
(3) Nothing is shown for `gr.Dataset` -- I assume that's just because the value is set to null, but flagging just in case:

image

(4) Uploading audio isn't working. I noticed the same for the Interactive Gallery. I think all Upload is broken.


https://github.com/user-attachments/assets/ee9c1267-15b5-46fd-aed3-0015a50b1cd3

**Comment by abidlabs** [comment]
(5) Issue #4 is preventing me from being able to fully test `gr.ImageEditor`, but it looks like the bottom of the component is cut off:

image

and also when I click on the draw tool, its very laggy and colors the canvas at the wrong spot when clicked

**Comment by abidlabs** [comment]
(6) Clicking on the Native Plots tabs causes the whole app to crash, with this appearing in the console:

image

Those are the issues I noticed, happy to give this another review afterwards as needed!

**Comment by pngwn** [comment]
Okay, _almost_ all of those are expected. The main purpose of this test app is to test the initial rendering (i.e. will the component initialise on the server) not the full functionality. This test app lacks the business logic of the gradio app to do that properly.

1. there is no i18n just an identity fn, so it is returning whatever is passed in, this is expected.
2. I have broken audio in several ways, working on a fix.
3. This is correct, the way we load examples is weird and the test app cannot do it.
4. Uploading doesn't work in the test app as there is no server, only a frontend, so we have nowhere to upload files _to_.
5. I broke the image height, also fixing.
6. This shouldn't happen, native plots should render anything in SSR.

So 1, 3, 4 are expected. I'll take care of the others.

For 1, 3, 4 specifically, as long as they work in gradio proper, then they are working as expected.

**Comment by pngwn** [comment]
Fixed everything but can't reproduce #4.


https://github.com/user-attachments/assets/a6d79268-40ac-4937-8aa3-f8f53065d4cd

**Comment by abidlabs** [comment]
> Fixed everything but can't reproduce https://github.com/gradio-app/gradio/pull/4.

Can't repro it anymore either!

**Comment by abidlabs** [comment]
Ok went through and tested everything with and without JS, everything works great @pngwn with one exception: native plots don't work _if_ you start with a different component and then switch to the native plots tab. If you hard refresh, then they work. Recording:


https://github.com/user-attachments/assets/6acb4fdc-5ff1-4bdb-82e8-800f0cee985c

**Comment by pngwn** [comment]
Oh interesting.

**Comment by pngwn** [comment]
Okay fixed!

Gonna merge this in so I can continue but feel free to make any additional comments and i'll fix in a follow up!

**Pull request merged**